packages-native/actions/src/camera/TakePictureAdvanced.ts

```diff
@@ -0,0 +1,213 @@
+// This file was generated by Mendix Modeler.
+//
+// WARNING: Only the following code will be retained when actions are regenerated:
+// - the code between BEGIN USER CODE and END USER CODE
+// Other code you write will be lost the next time you deploy the project.
+
+import { Alert, Linking } from "react-native";
+import ImagePicker from "react-native-image-picker";
+
+type PictureSource = "camera" | "imageLibrary" | "either";
+
+type PictureQuality = "original" | "low" | "medium" | "high" | "custom";
+
+/**
+ * @param {MxObject} picture - This field is required.
+ * @param {"NativeMobileResources.PictureSource.camera"|"NativeMobileResources.PictureSource.imageLibrary"|"NativeMobileResources.PictureSource.either"} pictureSource - Select a picture from the library or the camera. The default is to let the user decide.
+ * @param {"NativeMobileResources.PictureQuality.original"|"NativeMobileResources.PictureQuality.low"|"NativeMobileResources.PictureQuality.medium"|"NativeMobileResources.PictureQuality.high"|"NativeMobileResources.PictureQuality.custom"} pictureQuality - Set to empty to use default value 'medium'.
+ * @param {Big} maximumWidth - The picture will be scaled to this maximum pixel width, while maintaing the aspect ratio.
+ * @param {Big} maximumHeight - The picture will be scaled to this maximum pixel height, while maintaing the aspect ratio.
+ * @param {MxObject} pictureData - Additional info about the picture will be stored in this object. Create it before calling this action.
+ * @returns {Promise.<boolean>}
+ */
+export async function TakePicture(
+    picture?: mendix.lib.MxObject,
+    pictureSource?: PictureSource,
+    pictureQuality?: PictureQuality,
+    maximumWidth?: BigJs.Big,
+    maximumHeight?: BigJs.Big,
+    pictureData?: mendix.lib.MxObject
+): Promise<boolean> {
```

Comment:
Why not return the actual MXObject instead of boolean?

Reply:
In this way we can avoid the usage of object as parameter.. You can follow same approach as DeviceInfo

Reply:
I tried duplicating from GetDeviceInfo but got stuck. The action can also return when a promise is rejected because the user canceled. I did not get that but that could happen when the app has no privileges and the user choose not to grant them. I had a version where the MxObject had a PictureTaken flag that was set similar to the original result boolean. But it got messed up somewhere in the async calls and promises. The image got saved, the object was visible and filled in the debugger but was empty in the nanoflow. No idea why. That's why I switched to passing it in. Takes away the async create object. Maybe you have a better idea to make all the promises work together to return an object and still deal with the cancel exception?

Reply:
some promise was not being awaited I think.

Reply:
I still think the current behavior is not the expected one, we should avoid object parameter to be filled and return an object of an specific entity.

Reply:
I understand. I will try to refactor it, but that will not be for today. There is one catch with returning the object: Anonymous has no access rights to it. Best to keep it as a separate JS action then to not break any existing apps.

Reply:
I managed to make it work. I also found why the previous attempt did not work. If I set values on the result object before calling storeFile, they get lost on Android only, but not everytime. Now I fill the result object when storeFile is resolved, that works fine, on iOS and Android, both for camera and picking from the album.